Teach_Breakout.py

```diff
@@ -0,0 +1,344 @@
+    """Returns a copy of a keras model."""
+    model.save('tmp_model_x')
+    new_model = keras.models.load_model('tmp_model_x')
+    os.remove('tmp_model_x') # delete the model once it's been loaded. (Not clear if deleting the model is working correctly?)
+    return new_model
+
+
+# Modified slightly from the blog:
+def fit_batch_target(model, target_model, gamma, start_states, actions, rewards, next_states, is_terminal, number_in_batch):
+    """Do one deep Q learning iteration.
+
+    Params:
+    - model: The DQN
+    - target_model: The target DQN (copied from the DQN every e.g. 10k iterations)
+    - gamma: Discount factor (should be 0.99)
+    - start_states: numpy array of starting states
+    - actions: numpy array of actions corresponding to the start states (NOT one-hot encoded; normal integer encoding)
+    - rewards: numpy array of rewards corresponding to the start states and actions
+    - next_states: numpy array of the resulting states corresponding to the start states and actions
+    - is_terminal: numpy boolean array of whether the resulting state is terminal
+
+    """
+
+    # First, predict the Q values of the next states (using the target model):
+    next_Q_values = target_model.predict([next_states, np.ones((number_in_batch, 4))])
+    # The Q values of the terminal states is 0 by definition:
+    next_Q_values[is_terminal] = 0
+    # The Q values of each start state is the reward + gamma * the max next state Q value:
+    Q_values = rewards + gamma * np.max(next_Q_values, axis=1)
+    # TODO: Doublecheck: correct axis for max? And that next_Q_values gives 4 values (one for each action) for each item in the batch?
+
+    # Turn the actions into one-hot-encoded actions (required to use the mask)
+    one_hot_actions = np.zeros((number_in_batch, 4))
+    for i in range(number_in_batch):
+        for j in range(4):
+            if j == actions[i]:
+                one_hot_actions[i, j] = 1
+
+    # Fit the keras model, using the mask
+    model.fit([start_states, one_hot_actions], one_hot_actions * Q_values[:, None],
+        epochs=1, batch_size=len(start_states), verbose=0) # Should it be epochs=1??
+
+
+def make_n_fit_batch(model, target_model, gamma, iteration, mem_size, mem_frames, mem_actions, mem_rewards, mem_is_terminal, number_in_batch):
+    """Make a batch then use it to train the model"""
+
+    if iteration < mem_size:
+        # In this case, sample 'number_in_batch' numbers from 4 to 'iterations'
+        # (We use 4 so that there are 5 frames minimum, so we can still make 2 states (because each state is the frame
+        # plus the previous 3 frames))
+        indices_chosen = random.sample(range(4, iteration), number_in_batch)
+    else:
+        # Now the memory is full, so we can take any elements from it
+        indices_chosen = random.sample(range(0, mem_size), number_in_batch)
+
+    # Initialise the batches
+    start_states = np.zeros((number_in_batch, 105, 80, 4))  # (Array instead of list)
```

Comment:
Avoid magic numbers. Use `ATARI_SHAPE`. Note you can concatenate tuples like so:
`(numbers_in_batch, ) + ATARI_SHAPE`